bitcointx/core/key.py

```
@@ -1905,6 +2030,162 @@ def get_pubkey(self, key_id: bytes,
         return None
 
 
+T_XOnlyPubKey = TypeVar('T_XOnlyPubKey', bound='XOnlyPubKey')
+
+
+class XOnlyPubKey(bytes):
+    """An encapsulated X-Only public key"""
+
+    __fullyvalid: bool
```

I realise that this is not specific to the new taproot/BIP340 code (you already had it in `CPubKey`, but what is the significance of storing this `__fullyvalid` bool? Is it somehow useful to have a pubkey object that is from an invalid encoding?

---

- The Core makes that distinction for some reason
- It is used in tests to supply invalid pubkeys and check that the code fail with invalid pubkeys

My guess is that Core does this for speed so you can create `CPubKey` instance and check for validity later.

Maybe it makes sense to throw `ValueError` on invalid encoding and introduce `allow_invalid_encoding` kwarg for the tests ? This would be breakage of the API, but I doubt this breakage will affect much real-world code except the test code.

---

Address classes already have `accept_invalid` kwarg for `from_pubkey()` method, but I guess one could accidentally use invalid pubkeys in other places (like, constructing scripts) where they could miss checking the `is_fullyvalid()`.

---

OTOH, if we introduce that change, and then new code that uses bitcointx starts dropping `is_fullyvalid()` checks, but afterwards somehow is used with the old version of the lib, then it could pass invalid pubkeys where it should not. This is somewhat alleviated by the fact that the code may list the version requirement for the library, of course.

---

I think only two reasonable paths: leave as is, or completely remove, anything else would have the kind of problems you describe in that last comment. Even if I think the 'remove' option is preferable, the taproot patch is not really the right time to make such a big change.

> Address classes already have accept_invalid kwarg for from_pubkey() method

It's interesting that this option exists (even though of course default False). I suppose it might be related to burn addresses? (your performance hypothesis makes sense too, i guess). Generally I'd lean towards making no public API support for such things, since the risk of using them incorrectly outweighs the very small  amount of usage (burn addresses are very rarely used for anything).

But no action item here, imo: this is orthogonal to the taproot patch.

---

and that `test_from_invalid_pubkeys` was present in python-bitcoinlib since long time ago. So maybe this is really just an option to allow to create burn addresses.

---

> But no action item here, imo: this is orthogonal to the taproot patch.

I created a separate issue: https://github.com/Simplexum/python-bitcointx/issues/61